loki/transform/transform_lift_contained_subroutines.py

```diff
@@ -0,0 +1,212 @@
+        # 2. Variables in body.
+        # 3. Variable kinds.
+        # 4. Array dimension variables.
+        all_vars_body = FindVariables().visit(inner.body)
+        all_vars_spec = FindVariables().visit(inner.spec)
+        all_vars = set(var.clone() for var in all_vars_body.union(all_vars_spec))
+        kinds = set()
+        for var in all_vars:
+            if var.type.kind:
+                kinds.add(var.type.kind.clone())
+        all_vars = all_vars.union(kinds)
+
+        # Get the names of all global variables from the perspective of `inner`. 
+        gvar_names = set()
+        for var in all_vars:
+            # The next while loop is in case `var` is a member of a derived type. 
+            # For example for variable named `a%b%c`, 
+            # the global variable we want is `a`, not `a%b%c` itself.
+            varname = var.name.lower()
+            vartmp = var.clone()
+            while vartmp.parent is not None:
+                varname = vartmp.parent.name.lower()
+                vartmp = vartmp.parent
+            if not varname in lvar_names: # Only add variables that are not defined locally.
+                gvar_names.add(varname)
+
+        # `gvar_names` might still contain inline calls to functions (say, log or exp) 
+        # that show up as variables. Here they get removed.
+        inline_call_names = set([call.name.lower() for call in FindInlineCalls().visit(inner.body)])
+        gvar_names = gvar_names.difference(inline_call_names)
+
+        defs_to_add = [] # List of new arguments that should be added to the inner routine. 
+        imports_to_add = [] # List of imports that should be added to the inner routine. 
+        outer_spec_vars = FindVariables().visit(routine.spec) 
+        outer_imports = FindNodes(Import).visit(routine.spec)
+        outer_import_vars = FindVariables().visit(outer_imports)
+        outer_import_names = [var.name.lower() for var in outer_import_vars]
+        
+        gvar_names = list(gvar_names)
+        # This is the main loop that processes `gvar_names`, i.e decides how each global variable in 
+        # `inner` should be resolved.
+        # NOTE: It is very important to process `gvar_names` as a list (with an order) because the
+        # following loop may add new items that to `gvar_names` that need to be resolved. 
+        for gn in gvar_names:  
+            same_named_globals = [var.clone() for var in outer_spec_vars if var.name.lower() == gn]
+            if len(same_named_globals) != 1:
+                # If the code crashes to the next assert, there is a bug in the implementation of this function.
+                assert len(same_named_globals) == 0 
+                msg = f"Could not resolve the symbol '{gn}' "\
+                f"in the contained subroutine '{inner.name}'. "\
+                f"The symbol '{gn}' is undefined in the parent subroutine '{routine.name}'."
+                raise Exception(msg)
```

Comment:
Indeed it is. I went with `RuntimeError`. However, maybe there could be something like a custom error `SymbolResolutionError` or something. I'd imagine this kind of an error pops up in other places in Loki as well, or? I did not, however, implement this, because it is not strictly speaking part of this PR.